README.md
# Scrollable List

A New Flutter package which helps to scroll the content with their respective bottom bar menu for android and ios.

## Features

* Effective scroll with active tab indication

## Getting started

* Install this package for scrollable main contents with bottom bar

## Install

* Add the dependency

dependencies:
  scrollable_list: ^0.0.2+1

* Import the package

```dart
import 'package:scrollable_list/scrollable_list.dart';
```

## How to use

* Take list of bottom menus with title, icon and unique id.

* Write list of body contents with title , content and unique id.

* Make list as CommonList structure.

* Must use same unique id to relevant menu and page controls for scrolling purpose.

## Screenshots

<img src="/home.jpg" height="300em" />
<img src="/search.jpg" height="300em" />
<img src="/profile.jpg" height="300em" />
<img src="/settings.jpg" height="300em" />

## Example

```dart
class ScrollablePage extends StatelessWidget {
  final String title;
   ScrollablePage({Key? key,this.title = ""}) : super(key: key);
  /* Here we have the list of main content type CommonList, which is occupy the body of the page */ 
 /* Each common list has the label , content , icon and main widget */
    final list = <CommonList>[
       CommonList(uniqueID: 1,content: "Home page",label: "Home",mainWidget: const Content()),
       CommonList( uniqueID: 2,content: "Search page",label: "Search",mainWidget: const Content()),
       CommonList(uniqueID: 3,content: "Profile page",label: "Profile",mainWidget: const Content()),
       CommonList(uniqueID: 4,content: "Settings page",label: "Settings", mainWidget: const Content()),
     ];
      /*Here we have the list of bottom content type CommonList, which is occupy the bottom of the page */
     /* Able to add image element for each object */
  final bottomList = [
      CommonList(uniqueID: 1, content: "",label: "Home", image: "assets/images/home.png",icon: Icons.home ),
      CommonList(uniqueID: 2,content: "",label: "Search", icon: Icons.search_rounded,image: "assets/images/search.png" ),
      CommonList(uniqueID: 3,content: "",label: "Profile",height: 23,image: "assets/images/user.png", icon: Icons.person),
      CommonList( uniqueID: 4,content: "", label: "Settings",image: "assets/images/settings.png",icon: Icons.settings),
    ];
  final ScrollController controller = ScrollController();
  @override
  Widget build(BuildContext context) {
/* here we need to pass the body and bottom content to the page  */
    return SafeArea(
      child: ScrollableWidget(
        scrollController: controller,bodyContent: list,bottomContent: bottomList,
        appbar: CustomAppbar(title: title,),selectedColor: Colors.blue,unSelectedColor: Colors.black38,
      )
    );
  }
}
```
## Additional information

* This package has some required information for maintaining the scroll event with relevant bottom bar menu.

* This package uses the two list one is for body page content and another one is bottom bar content.

* List content must have the same unique id for relevant page and bottom menu element.

* Here, You can create own page design with contents,title and unique id  for all pages.

* This package supports four bottom bar menu and four page controls 

* It gives smooth scrolling and navigation.




CHANGELOG.md
## [0.0.1] -  Initial release

* A Customizable Scrollable list library written in dart.
* It helps to scroll content respective to specific bottom bar menu.

## [0.0.2] - Customized body widget

* Anyone can create main page design based on their needs.
* It supports four menu and smooth scrolling and navigation.

## [0.0.2+1] - Image parameter

* Able to add specific image for bottom bar menu
* Supports height parameter for image and size parameter for icons

lib/list.dart
import 'package:flutter/cupertino.dart';

/// Class represents the default structure with following parameters
class CommonList {
  String? label;
  String? content;
  IconData? icon;
  String image="";
  double height=24;
  Widget? mainWidget;
  int uniqueID = 0;

  CommonList(
      {this.label,
      this.content,
      this.image="",
        this.height=24,
      this.mainWidget,
      this.icon,
      required this.uniqueID});

  CommonList.fromJson(Map<String, dynamic> json) {
    label = json['label'];
    content = json['content'];
    uniqueID = json['id'];
    icon = json['icon'];
    height = json['height'];
    mainWidget = json['widget'];
    image = json['image'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['label'] = label;
    data['content'] = content;
    data['id'] = uniqueID;
    data['icon'] = icon;
    data['height'] = height;
    data['widget'] = mainWidget;
    data['image'] = image;
    return data;
  }
}


lib/appbar.dart
import 'package:flutter/material.dart';

class CustomAppbar extends StatelessWidget implements PreferredSizeWidget {
  /// Appbar widget with below elements
  final String leadingIcon;
  final String trailingIcon;
  final Function? leadingOnTap;
  final Function? trailingOnTap;
  final String title;
  final double height;
  final bool isPendingRequest;
  const CustomAppbar(
      {this.leadingIcon = "",
      this.trailingIcon = "",
      this.trailingOnTap,
      this.leadingOnTap,
      this.title = "",

      /// Initialization of height with value 30
      this.height = 30,

      /// Initialization of isPendingRequest with value false
      this.isPendingRequest = false,
      Key? key})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    /// Preferred size widget with title and icon children widgets
    return PreferredSize(
      preferredSize: preferredSize,
      child: SizedBox(
        width: MediaQuery.of(context).size.width,
        height: 90,
        child: Container(
          margin: const EdgeInsets.fromLTRB(0, 46, 0, 0),
          child: Padding(
            padding: const EdgeInsets.symmetric(horizontal: 15.0),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.start,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Expanded(

                    /// Top layout title
                    child: Text(
                  title,
                  style: const TextStyle(
                    color: Colors.black,
                    fontWeight: FontWeight.w500,
                    fontSize: 19,
                  ),
                )),

                /// Check whether trailingIcon is empty or not. if it is empty, SizedBox widget will be displayed
                trailingIcon.isEmpty
                    ? const SizedBox()
                    : InkWell(
                        onTap: () {
                          /// On tap event
                          trailingOnTap!();
                        },
                        child: Image(
                          image: AssetImage(trailingIcon),
                          fit: BoxFit.contain,
                          width: 30,
                          height: height,
                        ))
              ],
            ),
          ),
        ),
      ),
    );
  }

  /// Preferred size
  @override
  Size get preferredSize => const Size.fromHeight(60);
}


example/lib/content.dart
import 'package:flutter/material.dart';

class Content extends StatelessWidget {
  const Content({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const SizedBox(
          height: 10,
        ),
        const Text(
            "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.",
            style: TextStyle(fontSize: 15, height: 1.6)),
        const SizedBox(
          height: 20,
        ),
        const Text("Actions", style: TextStyle(fontSize: 18)),
        const SizedBox(
          height: 10,
        ),
        ListView.builder(
          shrinkWrap: true,
          scrollDirection: Axis.vertical,
          itemCount: 6,
          itemBuilder: (BuildContext context, int index) => Padding(
            padding: const EdgeInsets.only(bottom: 12.0),
            child: Container(
              decoration: BoxDecoration(
                  color: Colors.grey[50],
                  border: Border.all(color: Colors.grey[100]!),
                  borderRadius: BorderRadius.circular(10)),
              child: Padding(
                padding: const EdgeInsets.all(10.0),
                child: Row(
                  children: [
                    const Icon(
                      Icons.arrow_back_ios_rounded,
                      color: Colors.black38,
                      size: 20,
                    ),
                    const SizedBox(
                      width: 7,
                    ),
                    Text("Action ${index + 1} - You can Navigate through this",
                        style: const TextStyle(fontSize: 15, height: 1.3)),
                  ],
                ),
              ),
            ),
          ),
        ),
      ],
    );
  }
}


lib/scrollable_widget.dart
import 'package:flutter/material.dart';
import 'package:scrollable_list/appbar.dart';
import 'package:scrollable_list/list.dart';

class ScrollableWidget extends StatefulWidget {
  /// A stateful widget with below elements
  final PreferredSizeWidget appbar;
  final ScrollController scrollController;
  final List<CommonList> bodyContent;
  final List<CommonList> bottomContent;
  final bool isBottomLabelOnly;
  final Color selectedColor;
  final Color unSelectedColor;
  final bool isSelected;
  final int selectedIndex;
  final double titleFont;
  final double subTitleFont;
  final double iconSize;
  const ScrollableWidget(
      {Key? key,

      /// Initialization of customized app bar widget
      this.appbar = const CustomAppbar(
        title: "Scrollable List",
      ),
      required this.scrollController,
      required this.bodyContent,
      required this.bottomContent,
      this.isBottomLabelOnly = false,
      this.selectedColor = Colors.black,
      this.unSelectedColor = Colors.black26,
      this.isSelected = false,
      this.selectedIndex = 0,
        this.titleFont = 18,
        this.subTitleFont = 15,
        this.iconSize = 30})
      : super(key: key);

  @override
  ScrollableWidgetState createState() => ScrollableWidgetState();
}

class ScrollableWidgetState extends State<ScrollableWidget> {
  @override
  Widget build(BuildContext context) {
    /// Scaffold widget has app bar and notification listener as body widget.
    return Scaffold(
      appBar: widget.appbar,
      body: NotificationListener(
        /// Listener to observe the scroll event.
        child: ListView(
          scrollDirection: Axis.horizontal,
          controller: widget.scrollController,

          /// Map entries for body content list
          children: widget.bodyContent
              .map((e) => Padding(
                    padding: const EdgeInsets.symmetric(
                        horizontal: 13.0, vertical: 10),
                    child: Container(
                        color: Colors.white,
                        width: MediaQuery.of(context).size.width,
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.start,
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            const SizedBox(
                              height: 10,
                            ),
                            Text(e.content.toString(),
                                style:  TextStyle(fontSize: widget.titleFont)),
                            e.mainWidget!
                          ],
                        )),
                  ))
              .toList(),
        ),

        /// Function for listening scroll controls.
        onNotification: (t) {
          if (t is ScrollEndNotification) {}
          setState(() {
            if (MediaQuery.of(context).size.width <
                    widget.scrollController.position.pixels &&
                widget.scrollController.position.pixels <
                    MediaQuery.of(context).size.width * 2) {
              currentIndex = 1;
            } else if (MediaQuery.of(context).size.width * 2 <
                    widget.scrollController.position.pixels &&
                widget.scrollController.position.pixels <
                    MediaQuery.of(context).size.width * 3) {
              currentIndex = 2;
            } else if (MediaQuery.of(context).size.width * 3 <
                widget.scrollController.position.pixels) {
              currentIndex = 3;
            } else if (MediaQuery.of(context).size.width >
                widget.scrollController.position.pixels) {
              currentIndex = 0;
            }
          });

          // controller.currentindex.value = scrollController.position.pixels == 0.0?0:(scrollController.position.pixels>=21 && scrollController.position.pixels<=450 )?1:2
          return isExtended;
        },
      ),
      bottomNavigationBar: SizedBox(
        height: 60,
        child: Center(
          child: ListView(
            scrollDirection: Axis.horizontal,
            shrinkWrap: true,
            /// Map entries for bottom content list
            children: widget.bottomContent
                .asMap()
                .entries
                .map((e) => Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 20.0),
                      child: InkWell(
                        onTap: () {
                          setState(() {
                            // Assigning a index of map object to currentIndex variable.
                            currentIndex = e.key;
                          });

                          /// Scroll position change based on the current index.
                          widget.scrollController.animateTo(
                              (currentIndex.toDouble() *
                                      MediaQuery.of(context).size.width) +
                                  (currentIndex != 0 ? currentIndex * 20 : 0),
                              duration: const Duration(milliseconds: 1),
                              curve: Curves.fastLinearToSlowEaseIn);
                        },
                        child: Column(
                          children: [
                            /// Check whether the conditions is true/false , if it is true SizedBox will be displayed.
                            widget.isBottomLabelOnly
                                ? const SizedBox()
                                : e.value.image.isNotEmpty?Image.asset(e.value.image,height: e.value.height,
                              color: currentIndex == e.key
                                  ? widget.selectedColor
                                  : widget.unSelectedColor,
                            ):
                            Icon(
                                    e.value.icon,
                                    size: widget.iconSize,
                                    color: currentIndex == e.key
                                        ? widget.selectedColor
                                        : widget.unSelectedColor,
                                  ),
                              const SizedBox(height: 6,),
                            /// A title of the bottom bar menu
                            Text(
                              e.value.label.toString(),
                              style: TextStyle(
                                  fontSize: widget.subTitleFont,
                                  color: currentIndex == e.key
                                      ? widget.selectedColor
                                      : widget.unSelectedColor),
                            ),
                          ],
                        ),
                      ),
                    ))
                .toList(),
          ),
        ),
      ),
    );
  }

  /// Declaration of variables
  int currentIndex = 0;
  bool isExtended = false;
}


example/lib/scrollable_app.dart
import 'package:flutter/material.dart';
import 'package:scrollable_list/appbar.dart';
import 'package:scrollable_list/list.dart';
import 'package:scrollable_list/scrollable_list.dart';

import 'content.dart';

/// Class ScrollablePage with title argument
class ScrollablePage extends StatelessWidget {
  final String title;
  ScrollablePage({Key? key, this.title = ""}) : super(key: key);

  /// Unique id must be the same value in the below list
  /// A list and bottom list with uniqueID , content , label , mainWidget
  final list = <CommonList>[
    CommonList(
        uniqueID: 1,
        content: "Home page",
        label: "Home",
        mainWidget: const Content()),
    CommonList(
        uniqueID: 2,
        content: "Search page",
        label: "Search",
        mainWidget: const Content()),
    CommonList(
        uniqueID: 3,
        content: "Profile page",
        label: "Profile",
        mainWidget: const Content()),
    CommonList(
        uniqueID: 4,
        content: "Settings page",
        label: "Settings",
        mainWidget: const Content()),
  ];
  final bottomList = [
    CommonList(
      uniqueID: 1,
      content: "",
      label: "Home",
      image: "assets/images/home.png",
      icon: Icons.home,
    ),
    CommonList(
      uniqueID: 2,
      content: "",
      label: "Search",
      icon: Icons.search_rounded,
      // image: "assets/images/search.png",
    ),
    CommonList(
      uniqueID: 3,
      content: "",
      label: "Profile",
      height: 23,
      // image: "assets/images/user.png",
      icon: Icons.person,
    ),
    CommonList(
      uniqueID: 4,
      content: "",
      label: "Settings",
      // image: "assets/images/settings.png",
      icon: Icons.settings,
    ),
  ];
  final ScrollController controller = ScrollController();
  @override
  Widget build(BuildContext context) {
    return ScrollableWidget(
      scrollController: controller,
      bodyContent: list,
      bottomContent: bottomList,
      /// Customized app bar widget
      appbar: CustomAppbar(
        title: title,
      ),
      /// Size of the icon
      iconSize: 24,
      /// Color represents the selected element
      selectedColor: Colors.blue,
      /// Color represents the unselected element
      unSelectedColor: Colors.black38,
    );
  }
}
